### PR TITLE
Adding rate limiting to Admins.cshtml.cs

### PR DESCRIPTION
**Pull request by jrmccannon**
This adds rate limiting to the Admins page to address ability to spam emails using the invite admin form.

The settings used are 
```
            options.PermitLimit = 10;
            options.Window = TimeSpan.FromMinutes(1);
            options.QueueLimit = 0;
```

This means we limit 10 requests to the resource for 1 min. The `QueueLimit` determines the number of requests we'll hold in queue to execute later once the limit has been reached. This was set to 0 to drop all requests once the limit is reached.

I had to apply the rate limit attribute to the entire razor page due to a [limitation of the RateLimitMiddleware](https://learn.microsoft.com/en-us/aspnet/core/performance/rate-limit?view=aspnetcore-8.0#applying-attributes-to-razor-pages).

In order to also reduce the ability of a user to abuse this page, I enforced a unique restriction on the email the invite is being sent to.  If the email already exists in the invite list, we won't send the invite. They'll have to cancel the old invite first.

### CODE REVIEW

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
 using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.AspNetCore.RateLimiting;
 using Passwordless.AdminConsole.EventLog.Loggers;
 using Passwordless.AdminConsole.Helpers;
 using Passwordless.AdminConsole.Identity;
 using Passwordless.AdminConsole.Services;
 
 namespace Passwordless.AdminConsole.Pages.Organization;
 
+[EnableRateLimiting("fixed")]
```
Comment (jonashendrickx) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:14]: Remove the rate limiting on page level, instead move OnPostInvite to an endpoint, and have the form submission calling the endpoint.

Then apply the rate limiting to the endpoint.

`app.MapGet("/", () => Results.Ok($"Fixed Window Limiter {GetTicks()}"))
                           .RequireRateLimiting(fixedPolicy);`
                           
Consider also using a static readonly representation of the rate limiting policy.

**Review by jonashendrickx — CHANGES_REQUESTED** [review]

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs (hunk 1 above)
Comment (jrmccannon) [reply src/AdminConsole/Pages/Organization/Admins.cshtml.cs]: Chatted with @abergs about splitting this out into a separate API endpoint. He felt it wasn't worth moving it out of the page.  We decided upping the limit to 100 and put a few more safe guards in.  Only allowing 1 invite per email at a time and 10 pending invites at a time.

Also went ahead with the static definition of the policy and adding of rate limiting.

Comment (jonashendrickx) [reply src/AdminConsole/Pages/Organization/Admins.cshtml.cs]: Can you not use something better, i.e.:

1. Preventing double submissions in the first place
2. Upon generating links, I suppose they are valid for a limited time, so you can only have up to 3 active magic links for a given user or admin.

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 2]
```diff
@@ -92,28 +94,39 @@ public async Task<IActionResult> OnPostDelete(string userId)
...
     {
         CanInviteAdmin = await _dataService.CanInviteAdminAsync();
+
         if (CanInviteAdmin is false)
         {
             ModelState.AddModelError("error", "You need to upgrade to a paid organization to invite more admins.");
             return await OnGet();
         }
 
+        Models.Organization org = await _dataService.GetOrganizationAsync();
```
Comment (Tyrrrz) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:104]:
```suggestion
        var org = await _dataService.GetOrganizationAsync();
```

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 3]
```diff
@@ -92,28 +94,39 @@ public async Task<IActionResult> OnPostDelete(string userId)
...
+        var existingInvites = await _invitationService.GetInvitesAsync(org.Id);
+
+        if (existingInvites.Count == 10)
```
Comment (Tyrrrz) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:107]: Just in case

```suggestion
        if (existingInvites.Count >= 10)
```

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 4]
```diff
@@ -92,28 +94,39 @@ public async Task<IActionResult> OnPostDelete(string userId)
...
         if (!ModelState.IsValid)
         {
             // todo: Is there a pattern where we don't need to repeat this?
             return await OnGet();
         }
 
-        Models.Organization org = await _dataService.GetOrganizationAsync();
-        var orgId = org.Id;
-        var orgName = org.Name;
         ConsoleAdmin user = await _dataService.GetUserAsync();
```
Comment (Tyrrrz) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:125]:
```suggestion
        var user = await _dataService.GetUserAsync();
```

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 5]
```diff
@@ -92,28 +94,39 @@ public async Task<IActionResult> OnPostDelete(string userId)
...
+        Models.Organization org = await _dataService.GetOrganizationAsync();
+        var existingInvites = await _invitationService.GetInvitesAsync(org.Id);
+
+        if (existingInvites.Count == 10)
+        {
+            ModelState.AddModelError("error", "You can only have 10 pending invites at a time.");
+            return await OnGet();
+        }
+
+        if (existingInvites.Any(x => x.ToEmail == form.Email))
```
Comment (Tyrrrz) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:113]: Note that it might not catch all duplicates, including emails that use `+` or those written in a different case. If this is something that people may abuse, then maybe we need a more resilient check?

File: src/AdminConsole/RateLimiting/AdminPageRateLimit.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public const string PolicyName = "organizationIdFixedLength";
+    private const int PermitLimit = 100;
+    private static readonly TimeSpan Window = TimeSpan.FromMinutes(1);
+    private const int QueueLimit = 0;
+
+    public static RateLimiterOptions AddAdminPageRateLimitPolicy(this RateLimiterOptions limiter) =>
+        limiter.AddPolicy(PolicyName, context =>
+            RateLimitPartition.GetFixedWindowLimiter(
+                context.User.GetOrgId()!.Value.ToString(),
```
Comment (Tyrrrz) [inline src/AdminConsole/RateLimiting/AdminPageRateLimit.cs:17]: When can `.GetOrgId()` return null?

Comment (jrmccannon) [reply src/AdminConsole/RateLimiting/AdminPageRateLimit.cs:17]: If the user does not have the "OrgId" claim, which would occur if the user is unauthenticated (in which they would be redirected to the login page) or if they were never added to an org (which is very highly unlikely). 

TLDR: it shouldn't happen.

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs (hunk 5 above)
Comment (abergs) [reply src/AdminConsole/Pages/Organization/Admins.cshtml.cs:113]: good observation, in this case we just want a simple check. We should probably to a case-insensitive comparison.

**Review by abergs — APPROVED** [review]

**Comment by abergs** [comment]
Approved, see comments from Oleksii

**Review by abergs — DISMISSED** [review]

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 7]
```diff
@@ -92,28 +94,39 @@ public async Task<IActionResult> OnPostDelete(string userId)
...
+        var org = await _dataService.GetOrganizationAsync();
+        var existingInvites = await _invitationService.GetInvitesAsync(org.Id);
+
+        if (existingInvites.Count >= 10)
+        {
+            ModelState.AddModelError("error", "You can only have 10 pending invites at a time.");
+            return await OnGet();
+        }
+
+        if (existingInvites.Any(x => string.Equals(x.ToEmail, form.Email, StringComparison.CurrentCultureIgnoreCase)))
```
Comment (Tyrrrz) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:113]: I don't think non-ascii characters can be used in emails, can they?

```suggestion
        if (existingInvites.Any(x => string.Equals(x.ToEmail, form.Email, StringComparison.OrdinalIgnoreCase)))
```

File: src/AdminConsole/Pages/Organization/Admins.cshtml.cs [hunk 8]
```diff
@@ -92,28 +94,39 @@ public async Task<IActionResult> OnPostDelete(string userId)
...
-        ConsoleAdmin user = await _dataService.GetUserAsync();
-        var userEmail = user.Email;
-        var userName = user.Name;
+        var user = await _dataService.GetUserAsync();
 
-        await _invitationService.SendInviteAsync(form.Email, orgId, orgName, userEmail, userName);
+        await _invitationService.SendInviteAsync(form.Email, org.Id, org.Name, user.Email!, user.Name);
 
-        _eventLogger.LogInviteAdminEvent(user, form.Email, _systemClock.UtcNow.UtcDateTime);
+        _eventLogger.LogInviteAdminEvent(user, form.Email, _timeProvider.GetUtcNow().UtcDateTime);
```
Comment (Tyrrrz) [inline src/AdminConsole/Pages/Organization/Admins.cshtml.cs:129]: Nit/future: Maybe `TimeProvider` can be injected in some kind of context inside the event logger so that `_timeProvider.GetUtcNow().UtcDateTime` can be odne implicitly?

**Review by Tyrrrz — APPROVED** [review]

**Review by Tyrrrz — APPROVED** [review]
